src/rust/engine/fs/brfs/src/main.rs

```diff
@@ -0,0 +1,897 @@
+          Ok(Some(if is_executable {
+            executable_inode
+          } else {
+            non_executable_inode
+          }))
+        }
+        Ok(None) => Ok(None),
+        Err(err) => Err(err),
+      },
+    }
+  }
+
+  pub fn inode_for_directory(&mut self, digest: Digest) -> Result<Option<Inode>, String> {
+    match self.directory_inode_cache.entry(digest) {
+      Occupied(entry) => Ok(Some(*entry.get())),
+      Vacant(entry) => match self.store.load_directory(digest).wait() {
+        Ok(Some(_)) => {
+          let inode = self.next_inode;
+          self.next_inode += 1;
+          entry.insert(inode);
+          self.inode_digest_cache.insert(
+            inode,
+            InodeDetails {
+              digest: digest,
+              entry_type: EntryType::Directory,
+              is_executable: true,
+            },
+          );
+          Ok(Some(inode))
+        }
+        Ok(None) => Ok(None),
+        Err(err) => Err(err),
+      },
+    }
+  }
+
+  pub fn file_attr_for(&mut self, inode: Inode) -> Option<fuse::FileAttr> {
+    self.inode_digest_cache.get(&inode).map(|f| {
+      attr_for(
+        inode,
+        f.digest.1 as u64,
+        fuse::FileType::RegularFile,
+        if f.is_executable { 0o555 } else { 0o444 },
+      )
+    })
+  }
+
+  pub fn dir_attr_for(&mut self, digest: Digest) -> Result<fuse::FileAttr, i32> {
+    match self.directory_inode_cache.entry(digest) {
+      Vacant(entry) => match self.store.load_directory(digest).wait() {
+        Ok(Some(_)) => {
+          let inode = self.next_inode;
+          self.next_inode += 1;
+          entry.insert(inode);
+          self.inode_digest_cache.insert(
+            inode,
+            InodeDetails {
+              digest: digest,
+              entry_type: EntryType::Directory,
+              is_executable: true,
+            },
+          );
+          Ok(dir_attr_for(inode))
+        }
+        Ok(None) => Err(libc::ENOENT),
+        Err(err) => {
+          error!("Error loading directory {:?}: {}", digest, err);
+          Err(libc::EINVAL)
+        }
+      },
+      Occupied(entry) => Ok(dir_attr_for(*entry.get())),
+    }
+  }
+
+  pub fn readdir_entries(&mut self, inode: Inode) -> Result<Vec<ReaddirEntry>, i32> {
+    match inode {
+      ROOT => Ok(vec![
+        ReaddirEntry {
+          inode: ROOT,
+          kind: fuse::FileType::Directory,
+          name: OsString::from("."),
+        },
+        ReaddirEntry {
+          inode: ROOT,
+          kind: fuse::FileType::Directory,
+          name: OsString::from(".."),
+        },
+        ReaddirEntry {
+          inode: DIGEST_ROOT,
+          kind: fuse::FileType::Directory,
+          name: OsString::from("digest"),
+        },
+        ReaddirEntry {
+          inode: DIRECTORY_ROOT,
+          kind: fuse::FileType::Directory,
+          name: OsString::from("directory"),
+        },
+      ]),
+      // Skip digest contents because enumerating them takes forever.
```

Comment:
This would be "everything recursively visible within the `Directory` I have mounted?" One thing here is that we'll want to actually validate that a requested digests are members of the root `Directory` in order to prevent people escaping the sandbox that way. So eagerly listing the `Directory` might be a good idea anyway.

Reply:
Well, it's clear what's happening, but I don't think it's addressed my comment.... I probably left it in the wrong place.

My point was that you can escape the sandbox by requesting an explicit digest with `(DIGEST_ROOT, Some(digest_str))` that is not located under your root digest.

Reply:
Ok, I see what you mean. If a user process has access to this directory, then they have access to plenty of other things as well. But in the presence of a sandbox, you just wouldn't give them access to this directory.